### PR TITLE
Turns lightswitches off by default.

### PR DESCRIPTION
**Pull request by Apogee-dev**
Turns lightswitches off by default in empty departments, saving power and more importantly giving a visual indicator that a department is not staffed. Maybe now we'll spend less time standing in front of an empty RnD, cargo, or medbay hammering that bell.

EDIT: Since it's been brought up a few times, **this only affects areas with lightswitches.** Hallways, maint, and areas like medbay & brig lobby on most maps will not be affected.

:cl:
 * tweak: Lightswitches are now turned off by default.
 * tweak: Lightswitches check for players at roundstart and turn on if there are any present.

### CODE REVIEW

**Comment by Kurfursten** [comment]
This actually saves a non-negligible amount of power and will be very significant on lowpop in terms of lowering power demand. This doesn't mean the station will run out of power in its SMES slower though, since they don't adjust their output for load. It does mean that many rooms will run out of personal power slower (for example, unpopulated R&D would have working computers longer after power runs out in the SMES).

I estimate this will lower power grid load on most stations by about 56000W at shift start.

**Comment by Gazathor** [comment]
I disapprove of this for the singular fact that people might learn that light switches exist and I will lose a very easy way to fuck with people

**Comment by Sooose** [comment]
image
This has to be accounted for.

**Comment by Kammerjunk** [comment]
considering how few light switches are placed near room entrances as they probably should be, this sounds really annoying
let's do it

**Comment by Kurfursten** [comment]
We got those nice "glowy" lightswitches ages ago so it's not as bad as it sounds
Example, pitch black Tech Storage; can you find the switch?
image

**Comment by Apogee-dev** [comment]
> considering how few light switches are placed near room entrances as they probably should be, this sounds really annoying
> let's do it

Yeah this won't really come into its own unless the m*ppers take a pass with it in mind (Boxmed having no lightswitches for the central area & lobby comes to mind)

**Comment by smalIboss** [comment]
sounds cool and will make more people realize that light switches exist

**Comment by ghost** [comment]
merge now

**Comment by ShiftyRail** [comment]
Going to feel really awkward to spawn in an entirely dark brig or medbay.

**Comment by DamianX** [comment]
If it were up to me, I'd require:
- lights in a department to turn on automatically at roundstart if someone started the round in there
- changes to all maps to move light switches right next to the airlocks

before I'd consider merging this

**Comment by GreatBigFailure** [comment]
The switch should be turned on/turn on if a player spawns in their workspace round-start. 
I like the idea of knowing the manifest from the lights and like Kurf said, it'll be less no-power shuttle calls but I think it'll completely change the mood at round-start too. The central hallway lights should also be on by default.

**Comment by Apogee-dev** [comment]
>lights in a department to turn on automatically at roundstart if someone started the round in there
>The switch should be turned on/turn on if a player spawns in their workspace round-start.

A good idea and something I'll look into making happen. Not sure how to go about it, though- best guess, i need to dig into roundstart code.

>changes to all maps to move light switches right next to the airlocks

Honestly, needs to happen whether or not this gets merged- our switch placement sucks. Deserves its own separate PR, though.

>The central hallway lights should also be on by default.

This only affects rooms that come with a lightswitch at roundstart, so it never affects the central hallways unless a particularly sadistic mapper gives those switches.

**Comment by Dickwrath** [comment]
This doesn't sound friendly whatsoever to new players. The game is already hard enough to learn without being able to see dick on spawn. If you want more time to set up power, then make a PR adding a little more juice roundstart to the SMES's. Turning on the lightswitches every single round is going to get annoying.

**Comment by GreatBigFailure** [comment]
@Dickwrath Flipping switches on when someone spawns in an area is a good fix to that but it won't fix all issues like this, ya. Say a new player picks Chef for the first time. They walk into their walk-in freezer and the lights are completely off. New player won't be as likely to turn on the switch no matter how red and glowy it is and they won't know their PDA has a flashlight. Also there's a goat kicking them. 
A minor and very specific scenario but to be considered.

**Comment by drawsstuff** [comment]
I think this is really cool, atmospheric wise, but you may want to do what damian suggested

**Comment by kernikov** [comment]
> @Dickwrath Say a new player picks Chef for the first time. They walk into their walk-in freezer and the lights are completely off. New player won't be as likely to turn on the switch no matter how red and glowy it is and they won't know their PDA has a flashlight. Also there's a goat kicking them.

If this is the chef equivalent learning experience, like engineer forgetting to close the valve before removing a tank, so be it. If he dies, he dies.

**Comment by PrimeDSS13** [comment]
Will this affect rooms with no lightswitches that have lights on, such as maint rooms?

**Comment by Kurfursten** [comment]
It does not affect any room with no lightswitch, like medbay lobby, security lobby, or most maintenances.

**Comment by TGFH** [comment]
I'd be all for this if Damian's suggestions were implemented.

**Comment by Apogee-dev** [comment]
Implemented Damian's suggestions, or at least those I could without doing a bunch of mapping to improve switch locations. Lightswitches now turn on at roundstart if a player is detected in their area's parent department. This has a few weird side-effects like hydroponics being lit separately from bar & kitchen or the vacant office being lit if there's anybody in security, but as best as I can tell, this has to do with the maps themselves.

**Comment by Kammerjunk** [comment]
sweetness.
agree that light switches should be put next to doors and stuff, but also agree that it might be better suited for another PR

**Comment by Apogee-dev** [comment]
> agree that light switches should be put next to doors and stuff, but also agree that it might be better suited for another PR

I figure map changes are best left to map-focused PRs, especially if my hunch about area parenting is correct and any tweaks to that could be bundled in with changes to lightswitch placement.

**Comment by PrimeDSS13** [comment]
yeah might be for the best
excellent PR btw

**Comment by ghost** [comment]
Merge right now.

**Comment by Armadingus** [comment]
The nerf to dionas that no one saw coming.

**Comment by Eneocho** [comment]
> 
> 
> The nerf to dionas that no one saw coming.

Absolutely shattered my immulshon.

**Comment by Kurfursten** [comment]
I'm going to recuse myself from doing code review/merge on this PR because I literally wrote nearly all of the code used in it

**Comment by ghost** [comment]
> 
> 
> I'm going to recuse myself from doing code review/merge on this PR because I literally wrote nearly all of the code used in it

Golly, that sure is a weird thing to say given how often you self-merge your PRs.  Even this year alone.

#28771
#28770
#28768
#28700
#28687
#28673
#28581
#28492

**Comment by Kurfursten** [comment]
I had those looked over in code review by other collabs, there just weren't any changes to make. I may end up clicking the button on this too if someone else approves it. The only reason I made it clear here is because it's more obvious when I ping someone on Discord to say "Hey, review my PR" that I am asking for outside review.

If I ping another collab and ask them to review Apogee's PR with no explanation, it would look more like I'm handing out orders to review PRs instead of just doing code review myself. This all becomes a little more apparent when you look at the percentage of PRs that get reviewed and merged by me. In the last month alone, I merged 58 PRs.

File: code/__HELPERS/typeof.dm
Comment (ShiftyRail) [inline code/__HELPERS/typeof.dm]: You can actually replace this second `for` loop by 
```dm
out -= subtypesof(type)
```

As BYOND overrides the `list1 - list2` operator between lists to mean "remove everything in list2 from list1".

File: code/game/machinery/lightswitch.dm
Comment (ShiftyRail) [inline code/game/machinery/lightswitch.dm]: I could understand what you meant by this code with a little bit of thinking but it's not obvious. Could always add a comment explaining that you force the light switch to `off` by adding `0` as an argument.

Comment (Kurfursten) [reply code/game/machinery/lightswitch.dm]: another way you could make this obvious without a lengthy comment might be to include the arg name, e.g.: `toggle_switch(newstate = 0)` which I think is fairly obvious in its meaning

**Review by ShiftyRail — APPROVED** [review]

**Pull request merged**
